flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+
+        If the number of jobs present in the project can't be divided by the number
+        provided by the user then the last aggregate will have the maximum possible
+        jobs in it.
+        For instace, if 10 jobs are present in a project and they are aggregated in groups
+        of 3 then the length of aggregates would be 3, 3, 3, 1.
+
+        The below code-block provides an example on how jobs can be aggregated in
+        groups of 2.
+
+        .. code-block:: python
+
+            @aggregator.groupsof(num=2)
+            @FlowProject.operation
+            def foo(*jobs):
+                print(len(jobs))
+
+        :param num:
+            The default size of aggregates excluding the final aggregate.
+        :type num:
+            int
+        :param sort_by:
+            Before aggregating, sort the jobs by a given statepoint parameter.
+            The default behavior is no sorting.
+        :type sort_by:
+            str or NoneType
+        :param reverse_order:
+            States if the jobs are to be sorted in reverse order.
+            The default value is False.
+        :type reverse_order:
+            bool
+        :param select:
+            Condition for filtering individual jobs. This is passed as the
+            callable argument to `filter`.
+            The default behavior is no filtering.
+        :type select:
+            callable or NoneType
+        """
+        try:
+            if num != int(num):
+                raise ValueError('The num parameter should be an integer')
+            num = int(num)
+            if num <= 0:
+                raise ValueError('The num parameter should have a value greater than 0')
+        except TypeError:
+            raise TypeError('The num parameter should be an integer')
+
+        def aggregator_function(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        aggregator_obj = cls(aggregator_function, sort_by, reverse_order,
+                             select, f'groupsof{num}')
+
+        if num == 1 and sort_by == select is None and not reverse_order:
+            aggregator_obj._is_aggregate = False
+
+        return aggregator_obj
+
+    @classmethod
+    def groupby(cls, key, default=None, sort_by=None, reverse_order=False, select=None):
+        """Aggregates jobs according to matching state point key values.
+
+        The below code-block provides an example on how to aggregate jobs having
+        common state point parameter 'sp' whose value, when not found, is replaced by a
+        default value of -1.
+
+        .. code-block:: python
+
+            @aggregator.groupby(key='sp', default=-1)
+            @FlowProject.operation
+            def foo(*jobs):
+                print(len(jobs))
+
+        :param key:
+            Parameter which specifies how to group jobs based on a single
+            or a list of state point parameters or a custom callable.
```

Comment:
```suggestion
            The method by which jobs are grouped. It may be a state point
            or a sequence of state points to group by specific state point
            keys. It may also be an arbitrary callable of :class:`signac.Job`
            when greater flexibility is needed.
```